### DISCUSSION TITLE
Hans Discussions

### DISCUSSION THREAD

**Discussion by nordseele** [nordseele/hans#10] [General]
...

**Comment by waxcorp** [comment]
Greetings! I'm interested to learn what the current state of affairs is around the project? From what I can tell this might be the furthest along example of an open source MIDI -> i2c solution for the ER-301 after having sifted through the forums for the past few days. I wanted to find some existing Eurorack hardware that might be able to host something like hans so I just ordered a pre built Europi which seems like it might be enough to get things running? Thanks for your help and for all the hard work, and happy new year!

**Comment by nordseele** [comment]
Hi and thanks for your message! 

Short story:
Well, this project was not permanently cancelled... but it has been at a standstill for a very long time now and its future in very uncertain... :-/
Sorry, but the code on this GitHub repo will not work on _Europi_  unfortunately. I had to check what this module is but from what I' found, the module is using a RPI Pico, totally different hardware than what's required for Hans.
The code here is written in Rust and is compiled for Raspberry PI Zero W.

More infos below.


Longer story:
I should mention the status of this project on the Github page to avoid disappointments. There was a project for a version 2 running on RPI Pico W -much smaller, powered from 12V etc - but I never found time to program its firmware and Rust embedded is not yet fully working on Pico W. I have the prototype but no finished code for it. It is not a straight-forward move to port this version on RPI Pico. Rust is not yet fully compatible with Pico W. I started coding a V2 in C a few months ago, but stopped for many reasons- the main one is that I'm not personally using the ER-301 and other Eurorack modules anymore, the motivation has clearly faded for this specific reason, but there's also the impossibility to find some Pico W, components out of stock etc... Not a good time to run such a niche project imo. 

Hans was made for OSC in the first place. Midi was added later but never really finished. The reason is that there never was a consensus around a universal MIDI mapping for the ER-301. (Midi to ii via i2c) 

I asked on Lines, on O/D forum etc. So to use MIDI to i2c, the end-user would have to create a mapping and recompile the code, not very user friendly... Orthogonal Devices doesn't want to add a MIDI module ( at least CC and  Note) to the ER-301, so it is very complicated to create a universal mapping (The MIDI protocol is comprised of infos between 0 and 127 (notes, velocity, aftertouch, continuous controller etc. and on the other hand the ER-301 has 100 SC.CV receivers and 100 SC.TR receivers.) I understand and tend to agree -now- with Brian's position on MIDI & Eurorack though. That being said, the technical system for converting MIDI to "ii" (protocol understood by the ER-301) do exist in the code, but at the moment, if Hans receives a CC or MIDI note, it will not transmit anything to the 301, not knowing what to do with it. What this code does "out of the box" is mainly OSC to i2c ( OSC with variable latency, depending on the router etc so not optimal for time-sensitive things). I think it's been two years since I used the ER-301, to be honest, I have no idea at the moment if a version 2 will ever happen.

The code on this Github page **requires a Raspberry PI Zero W** and the add-on board or "hat" - see below. It will NOT work on a Raspberry Pi Pico as is and from what I see, Europi seems to use a Pico, so you should NOT expect this code to work on it, Pico is totally different hardware than a larger RPI like the Zero W running Linux. You might ask on the [Lines forum](https://llllllll.co/t/i2c2midi-a-diy-module-that-translates-i2c-to-midi/40950), I think there's a module using a Teensy that does i2c to MIDI, maybe it does Midi to i2c too and has compatibility with the ER-301 now. If it has a circuit with pull-up resistors, it should be able to communicate with the 301, but I have no idea if the creator of this projet has made something for the "Sound Computer".

Hans add-on board / hat -> Hans was also the name of the add-on board including the i2c circuit (2 pull-up resistors) and the MIDI circuit as well as a 5V circuit with fuse protection. This must be taken into account. I remember having a lot of problems with the 301 connected directly to the GPIO "hardware i2c" of a Raspberry Pi. The Er-301 did not start, I had to add these "pull-up resistors" so that the PI and the module communicate easily. I have a prototype of version 2 for RPI pico in stock, but it doesn't have a firmware as mentioned earlier, in other words it's pretty useless; I no longer have PCBs or prototypes of version 1 (for Raspberry PI Zero W) and on top of that I recently checked and some components I used, like the opto-coupler of the MIDI circuit were out of stock until September 2023. I'm not planning on having V1 boards made again and I don't have them in stock anymore ; I'm not planning on open-sourcing the full V1 boards but probably the schematics which are quite simple.

Of course, I'll post an update on this GitHub repo if things change. 

Best,

**Reply by waxcorp** [reply]
All makes sense and thanks very much for the super fast and detailed reply.

I'll let you know if my tinkering with the Pico and your Rust libraries bear any success and I intend to publish my scribblings on GitHub if so also.

Also appreciate the note about the OSC vs MIDI situation. I've been thinking of attempting some kind of a ChucK OSC -> OSC/i2c bridge -> ER-301 so that might be one thing to try with the Rust lib. I have a Percussa SSP which has an i2c bus that's gathering dust because the Percussa OS is so unusable, so the SSP hardware might be an interesting platform for such an experimental cast of characters. If I can get ChucK running on the SSP armhf architecture and enable the i2c bus it will all be loopback UDP traffic so hopefully the latency won't be noticeable. I'm looking for something well below 1ms whatever avenue I go down, which may mean I should instead go hunt down a Teensy as you referenced.

Another thing that seems to have materialized a year or three ago is the 16n MIDI -> i2c -> ER-301 path. If that works it would be a nice benefit to add a 16n to my rig but doing stuff with a Linux arm machine and your Rust lib seems more flexible and interesting if I don't hit some technical wall with the hardware. We'll see. Thanks again and happy new year!

**Comment by nordseele** [comment]
Yes, sure, please let me know!
If you have a Rpi Zero W or even a RPI 3, you could also try this Rust code by connecting the Pi directly to the ER-301 but again there's the issue (see below) with the "hardware i2c" bus of the RPI and a conflict with the 301 (+ it's always risky to connect the GPIO of a RPI to an expensive Eurorack module like that...) I remember that both had to start simultaneously or the 301 wouldn't boot at all (scary)...
I documented my findings and the possible cause of the issue in another repo in 2020. For what it's worth, I've just made a copy of the page here, in this public repo https://github.com/nordseele/hans/blob/master/docs/legacy_notes.md

I had to create a "software i2c" bus and to add 4.7k pull-up resistors (schematics on the same page)... It's a shame, because without this i2c bus issue, Hans/ a RPI wouldn't even need an external circuit etc. You probably won't encounter the problem with a Pico or other device, as long as it doesn't "acknowledge" or reply to anything sent by the ER-301 during its boot sequence but they'll also need some kind of P-U resistors on the i2c bus, there ain't any on the 301 side.

I'm not familiar with the SSP hardware but I know the hardware of the 16n quite well and all the modules borrowing or at the origin of the "ii" protocol, Txo, Teletype, Crow. Speaking of which, for MIDI to i2c you should also ask the creator of i2ctoMidi module (or monitor the thread on Lines) if he has plan to make it talk to the 301 too. His module was doing i2c from Teletype to Midi in the first place, at the same time he released it I was writing some C code doing the same thing for someone, but running on RPI. Anyway, I think he recently got a 301, and the codebase is actively maintained (also by Scanner Darkly), and the hardware should be capable of talking to the 301 or TXo etc... Of course there will always be the "MIDI mapping" that needs to be determined at some point ^^

Happy new year to you too ;-)

P.S: the infos provided on the "legacy notes" are probably not up to date and the steps provided are not to be followed as is, they're from a time Hans was made with Node JS, and the RPI config file might not be up to date either. But at least they give an idea of the schematic for the i2c bus etc
